### DISCUSSION TITLE
GDB not working correctly?

### DISCUSSION THREAD

**Discussion by osteffenrh** [coconut-svsm/svsm#215] [General]
I am unable to resolve symbols in the back-trace in gdb and can't list code at current position.

Is it me/my setup or is the debugging functionality with the gdb-stub not working correctly at the moment?

Here is what I did on current main branch (https://github.com/coconut-svsm/svsm/commit/4f312a3c2457595152dab568431fe52e2e566c87)

```
diff --git a/src/svsm.rs b/src/svsm.rs
index 40fa2a9..10f95c8 100755
--- a/src/svsm.rs
+++ b/src/svsm.rs
@@ -440,6 +440,7 @@ pub extern "C" fn svsm_main() {
         }
     }

+    debug_break();
     guest_request_driver_init();

     if let Some(ref fw_meta) = fw_metadata {
```

```
FEATURES=enable-gdb make
```

Run in Qemu as usual:
```
...
[SVSM] ***********************************
[SVSM] * Waiting for connection from GDB *
[SVSM] ***********************************
[SVSM] Unknown command: Ok("vMustReplyEmpty")
[SVSM] Unknown command: Ok("!")
[SVSM] Unknown command: Ok("qTStatus")
[SVSM] Unknown command: Ok("qSymbol::")
```

Connect gdb:
```
$ gdb --ex "target extended-remote /dev/pts/4"
GNU gdb (Fedora Linux) 14.1-1.fc39
[...]
(gdb) symbol-file target/x86_64-unknown-none/debug/svsm
Reading symbols from target/x86_64-unknown-none/debug/svsm...
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0xffffff8000000487 in switch_context ()
#2  0x0000000000000000 in ?? ()
(gdb) l
warning: 1	/rust/deps/compiler_builtins-0.1.103/src/lib.rs/@/compiler_builtins.b29fb1ecf5adec5f-cgu.018: No such file or directory
```

```
$ rustc -V
rustc 1.75.0 (82e1608df 2023-12-21)
$ which rustc
~/.cargo/bin/rustc
```

I checked a few commits from withing the last few weeks and they all behave the same.
I found a random commit where it still works: 649c6380e981bd90f33bc3c92ee7b513174a747f
I did not biscect this completely though.

Any hints welcome!

Thanks!

**Comment by roy-hopkins** [comment]
I've just tried this myself and see the same issue. 

Interestingly though, if you enable the breakpoint by removing the comment on the `debug_break()` near the start of `svsm_main()` instead of inserting it at the line you've added it on then it works correctly. You can then set a breakpoint in GDB on the call to `guest_request_driver_init()` and continue. Everything seems to work OK from that point.

Can you try that out to see if you same the behaviour? If so that can serve as a workaround for now whilst I investigate the issue.

Thanks!

**Comment by roy-hopkins** [comment]
Ok, I found the problem. When the first breakpoint is hit after a number of tasks have been created, which occurs when `start_secondary_cpus()` is called in this case, then the remote GDB obtains incorrect information about which thread is currently active.

You can verify this by reproducing the issue as you saw it but then issuing the command:

```
(gdb) info threads
  Id   Target Id            Frame 
* 1    Thread 1.5 (Running) 0x0000000000000000 in ?? ()
  2    Thread 1.4 (Running) 0x0000000000000000 in ?? ()
  3    Thread 1.3 (Running) 0x0000000000000000 in ?? ()
  4    Thread 1.2 (Stopped) 0xffffff8000096630 in svsm::debug::gdbstub::svsm_gdbstub::debug_break () at src/debug/gdbstub.rs:172
```
You can see that the task with the ID 5 (1.5 in the list above) is assumed to be the active thread. BTW, the text "(Running)" and "(Stopped)" comes from the stub itself and has no effect on the remote GDB apart from being reported in this list.

If you issue the command below then you can continue debugging as normal and see the backtrace correctly:

```
(gdb) thread 4
[Switching to thread 4 (Thread 1.2)]
#0  0xffffff8000096630 in svsm::debug::gdbstub::svsm_gdbstub::debug_break () at src/debug/gdbstub.rs:172
172	                asm!("int3");
(gdb) bt
#0  0xffffff8000096630 in svsm::debug::gdbstub::svsm_gdbstub::debug_break () at src/debug/gdbstub.rs:172
#1  0xffffff8000003e5b in svsm::svsm_main () at src/svsm.rs:445
#2  0xffffff80000858c0 in ?? ()
```
The root cause was down to the order the tasks were reported from the GDB stub. When the remote GDB is initially connected, the first task list report must place the current task at the start of the list. This was not the case. I've created a PR to fix this: #216.

**Reply by osteffenrh** [reply]
Thanks @roy-hopkins, for this workaround and for the fix. :-)
